### ISSUE TITLE
Allow seqan to be a dynamic library?

### ISSUE THREAD

**Issue by mr-c** [seqan/seqan#2291]
https://vittorioromeo.info/index/blog/2016_cpp_library_configuration_api.html

This would be great for Debian as it would speed up our builds and allow us to fix bugs and security issues faster.

Once multiple code paths for different processors are figured out then all seqan using apps would automatically benefit

However I am not an expert in this area at all, so I defer to the Seqan team's judgement here

Cheers,

**Comment by marehr** [comment]
Hi mr-c,

thank you for your input. We had some ideas in that direction.

We had a couple of discussions regard this, because seqan takes for ever to build and some modules blow up to a couple of gigabyte memory consumption.

We have some  major problems:
* Almost all our functions use templates and those can't be put into a linkable `.cpp` file. We rarely fully define a type, function or class that we could put that into a `.cpp`.
* We have a problematic include policy. As a developer when you change one file you will most-likely recompile all of the test cases instead of only those that are really affected, because we cross include almost everything of the library into each module. This is a real mess. One of the goals in seqan3 is to mitigate that.
* We tried PCH, but it does not work as we thought it would. I do not remember, but I think one of the above problems was a factor.
* We want the code to be as efficient as possible and let the compiler inline as much as possible, which he can only do if he can see the code. I know that there are some (I think it was called LTO) techniques to optimize during/after linking.

I'm not sure if something like this exists, but it would be cool to fully define some commonly used types, s.a. `using int_string = seqan::String<int>` that instead of recompiling them every time would be directly used from a `.so` or so.

Our library is similar to the STL where you can't make everything shared.

**Comment by h-2** [comment]
I agree with @marehr's points. It is not possible in general, because SeqAn is a template library and not a regular library. Instantiating all *relevant* codepaths (which wouldn't even be all possible codepaths) would likely result in a shared object file of over hundreds of MBs. Applications like Lambda that instantiate a relevant, but not exhaustive part of SeqAn weere already over 30MB binary size before I started cutting down on certain options to prevent over-instantiation. Shipping and linking this file would not only be difficult, it would also increase memory consumption of all the apps and likely decrease performance, because of broken cache-coherency and huge memory jumps within the shared library.

It might make sense to build a certain limited subset of SeqAn as a shared library so that other programming languages can make use of SeqAn functionality (Python, Rust, Go...), but this would always only be a very limited subset and it would need to be maintained and documented as a subset which in itself would require a substantial amount of work that currently no-one has time for I am afraid [Although I think that this would be a very worthwhile project].

**Comment by marehr** [comment]
@h-2, @mr-c, @rrahn 
Okay, it bothered me and I tried to look at it and I found a solution to

>  I'm not sure if something like this exists, but it would be cool to fully define some commonly used types, s.a. `using int_string = seqan::String<int>` that instead of recompiling them every time would be directly used from a `.so` or so.

With this we could add a shared library for common template instantiations like `seqan::String<int>`. This of course comes with a possible impact on the runtime, but now apps could share the same code which would drastically reduce the compile and binary size. Which I guess is more important for package maintainers.

For an example look at https://github.com/marehr/cpp_template_external_linkage_test (it is actually the same technique which the standard library uses)

From my POV it seems as this approach seems feasible and could be implemented into SeqAn as it is.

**Comment by h-2** [comment]
We will reconsider this for SeqAn3
